### PR TITLE
Update buildSpagoLock and refactor

### LINKED ISSUE #68: Set meta.mainProgram, use lib.getExe
Apps are using something like `bin/${bin.pname or bin.name}`. This is a bit hacky/brittle, but happens to work in all current cases. There’s already a `meta` place to set the main program--which can be used in tandem with `lib.getExe` (which calls `mainProgram` from `$src/bin`). There wouldn’t be a huge code reduction or anything, but would follow more expectations & could help downstream users, especially if not familiar with the tooling. If projects upstream in `nixpkgs` need their `meta.mainProgram` set, I can look into that.

### PR DESCRIPTION
**Pull request by deemp**
Related to https://github.com/purescript/spago/pull/981.

- Update `buildSpagoLock`
- Set `mainProgram` in packages so that `lib.getExe` works correctly (closes #68).
- Update the `generate` package
  - Update registry version
  - Update code
- Refactor `flake.nix`
- Set `alejandra` as the default formatter
  - Run `nix fmt` to format Nix files

### CODE REVIEW

**Comment by deemp** [comment]
@toastal, could you please review? IDK why garnix is unhappy. `nix flake check` works on my Linux.

File: flake.nix
Comment (toastal) [inline flake.nix]: Doesn’t the `{ xx, ... }@inputs` need to come after? Regardless, long-term these are easier to maintain as just calling `inputs.spago` in the one places you needed.

File: flake.nix
Comment (toastal) [inline flake.nix]: You want to avoid using `rec`

File: generate/default.nix
Comment (toastal) [inline generate/default.nix]: Should this be `ln -s` too?

File: nix/build-spago-lock.nix
Comment (toastal) [inline nix/build-spago-lock.nix]: same about `ln -s`. There are some trade offs, but I think it saves disk space more generally & an occasional optimization of the storage can clean up loose links.

File: spago-dev.nix
Comment (toastal) [inline spago-dev.nix]: Git is needed!?

**Comment by toastal** [comment]
Generally things look okay, but I’m not entirely sure on much of the context here. So I can’t really comment on the larger picture here. One thing that _feels_ like something that could be looked at is shrinking the `src`s with `lib.fileset` <https://nix.dev/tutorials/file-sets.html> which would make the derivations smaller by removing files that aren’t used, but I can’t really tell where exactly--just spider sense.

File: spago-dev.nix
Comment (deemp) [reply spago-dev.nix]: Yes. From the spago [README](https://github.com/purescript/spago#design-goals-and-reasons):

> Minimal dependencies: users should not be expected to install a myriad of tools on their system to support various workflows. Spago only expects git and purs to be installed.

**Comment by deemp** [comment]
> One thing that feels like something that could be looked at is shrinking the srcs with lib.fileset

Yes, I had the same idea about `lib.fileset`.

However, I don't know whether it's possible to determine a "working set" of files that are necessary to build a package. Consider a case when a fileset is determined based on a local project directory. If that project contains symlinks to files located outside of that directory, symlinked files won't be included into the derivation `src`. 

Thus, I think we should tell users to pass to `buildSpagoLock` an `src` built via `lib.fileset` so that:

- they ensure all necessary paths are present
- they have smaller derivations
- they rebuild their derivations less often due to ignoring some irrelevant paths in `./.`

File: nix/build-spago-lock.nix
Comment (deemp) [reply nix/build-spago-lock.nix]: AFAIU, it's only possible to symlink something in a derivation `$out`.

Hence, to cache the built `output`, it's necessary to copy it to `$out` at least once.

File: generate/default.nix
Comment (deemp) [reply generate/default.nix]: Yes. Fixed in https://github.com/thomashoneyman/purescript-overlay/pull/71/commits/163833707840fbf6fd0511be990bdee81438a4ea

File: flake.nix
Comment (deemp) [reply flake.nix]: Sure. https://github.com/thomashoneyman/purescript-overlay/pull/71/commits/6d5c9732e6d2edf1f7771b5c06e8e8a0c93b29b1 doesn't introduce `rec` anymore.

File: flake.nix
Comment (deemp) [reply flake.nix]: It can be before or after arguments, according to [docs](https://nixos.org/manual/nix/stable/language/constructs.html#functions). Anyway, https://github.com/thomashoneyman/purescript-overlay/commit/6d5c9732e6d2edf1f7771b5c06e8e8a0c93b29b1 doesn't introduce it anymore

**Comment by toastal** [comment]
Ah ha. That makes sense, especially with better context of how this is actually working now.

**Comment by thomashoneyman** [comment]
Hey @deemp! I’m happy to optimize the buildSpagoLock stuff, and we can use Spagonto build projects now that it has a working offline mode and the new lock file format.

However, I don’t want to build Spago from source if we can stick with the current approach of grabbing “binaries” from NPM. Would you mind explaining the motivation to build from source here? To be blunt it will make for a much faster review on my part if we narrowed to just changes to how buildSpagoLock etc works.

I apologize for slow responses this week, it’s a hectic one!

**Comment by deemp** [comment]
Hello, @thomashoneyman!

> Would you mind explaining the motivation to build from source here?

My goal is to be able to incrementally package PureScript projects using Nix.
I want to separate a derivation for a package and a derivation for its dependencies.

I build spago here because:
- I can debug the `spago` derivation and `buildSpagoLock` in the same flake
- I can check the PR against CI

> However, I don’t want to build Spago from source if we can stick with the current approach of grabbing “binaries” from NPM. 

I think it's not necessary to build or provide the `spago-dev`  (built from source) derivation here.
So, I can remove the commit that adds the spago derivation.

However, I think it may be convenient to have a derivation in the `spago` repo so that people can access the dev version without having to wait for a new release and an update of `purescript-overlay`.

I see that the `spago` repo uses `purescript-overlay` as an input ([link](https://github.com/purescript/spago/blob/bbe37b6cd497aa544bd0761fa7a56a5f5d002a87/flake.nix#L4)). Hence, if `buildSpagoLock` can build a package using the `spago` executable in offline mode, the `spago` repo can provide a derivation at almost zero cost.

> To be blunt it will make for a much faster review on my part if we narrowed to just changes to how buildSpagoLock etc works.

Currently,
- `buildSpagoPackages` builds given packages because AFAIU spago can't build dependencies only (https://github.com/purescript/spago/issues/1181)
- `buildSpagoLock` builds a given package given `.spago` and writable `output` of the package. 

Now, I see that my code doesn't account for `~/.cache/spago-nodejs` as `dream2nix` does ([link](https://github.com/nix-community/dream2nix/blob/16f2c654c6673a2b72225c81fb73af830139e33c/modules/dream2nix/WIP-spago/default.nix#L55)).

- Would you like to finish `buildSpagoLock`? If no, I'd like to ask several other questions.

- Is interaction with that directory still necessary?
- Where can I find the docs on it?
- Why is garnix unhappy?
- What else should be considered?

**Comment by deemp** [comment]
I've just realized I won't be able to build each package separately because `spago` wants `~/.cache/spago-nodejs` and will insert registry entries for each package into an sqlite database (it'll take too long).

I can build all dependencies at once, but it's less granular than the current solution that uses just the `purs` compiler.

So, I'd rather fix the generated `BuildInfo.purs` format.

File: flake.nix
Comment (toastal) [inline flake.nix]: What is the point of this? Do you need a specific feature from 20? Even if you did, you should do the check to see if it’s the current version is already greater as this will be unnecessary in the future or fail when v20 is deprecated.

File: generate/bin/src/Run.purs
Comment (toastal) [inline generate/bin/src/Run.purs]: Shouldn’t these be like `Console.error`?

Comment (deemp) [reply generate/bin/src/Run.purs]: IDK. @thomashoneyman?

BTW, there are quite many

```purs
Console.log "something"
liftEffect $ Process.exit' 1
```

It may be better to use a dedicated function.

Comment (thomashoneyman) [reply generate/bin/src/Run.purs]: Sure, it probably ought to be `Console.error`; the way this is run all the output is going to the console so there won't be an observable difference, but it's the proper thing to do.

@deemp we could add a `die` function that wraps this up, I'm happy with that. It's what Spago does.

**Comment by thomashoneyman** [comment]
I'm sorry I haven't looked at this yet — I'm having an exceptionally busy work + life period at the moment.

File: spago-dev.nix
Comment (thomashoneyman) [reply spago-dev.nix]: It's needed because Spago uses git to interact with the registry repositories. If we only use spago in offline mode (ie. `spago build --offline`) then I believe the calls to fetch the registry will be skipped and we don't need `git`.

File: flake.nix
Comment (deemp) [reply flake.nix]: fixed in https://github.com/thomashoneyman/purescript-overlay/pull/71/commits/2b33591a4e03103de685d84cad6cd26788d0927d

**Comment by bakhtiyarneyman** [comment]
@deemp Thank you for your contributions! I'm excitedly waiting to see it land to be able to send my own PRs. 

I believe the maintainers of this repo are using alejandra for formatting (although I see a few exceptions here and there). I think you are using a different formatter? Is it possible for you to reformat the code you touched with alejandra?

File: generate/bin/src/Run.purs
Comment (deemp) [reply generate/bin/src/Run.purs]: applied the suggestion in https://github.com/thomashoneyman/purescript-overlay/pull/71/commits/62ce26e10049f535f9acda09ca68332df888ea77

**Comment by toastal** [comment]
A check could be added like in <https://github.com/purescript/registry-dev/pull/685/commits/bc61b984afcad3a75227625f0107123157cdabe3>, but oof does alejandra have the worst of the formats IMO vs. nixpkgs-fmt or nixfmt. Upstream in Nixpkgs they are looking to make modifications to nixfmt to look more like nixpkgs-fmt, but that hasn’t yet landed
-- 
toastal ไข่ดาว | https://toast.al
PGP: 7944 74b7 d236 dab9 c9ef  e7f9 5cce 6f14 66d4 7c9e

**Comment by deemp** [comment]
>  I believe the maintainers of this repo are using alejandra for formatting (although I see a few exceptions here and there).

@bakhtiyarneyman, I preferred nixpkgs-fmt because of https://github.com/kamadorueda/alejandra/issues/372. However, as that issue is resolved now, I should check alejandra again.

In this repo, I initially used my hands :tm:  to not break the surrounding formatted code. I noticed that mostly only my code was re-formatted after I added alejandra as the default formatter and formatted the code. Hence, I think you're right that maintainers used alejandra.

**Comment by deemp** [comment]
> A check could be added like in <https://github.com/purescript/registry-dev/pull/685/commits/bc61b984afcad3a75227625f0107123157cdabe3>

Neat trick!

**Comment by thomashoneyman** [comment]
For what it’s worth I don’t have an opinion on what formatter is used so long as we use one. If it’s easier for yall to use a different one then I’m fine with that.

**Comment by deemp** [comment]
The default formatter for now is alejandra.

One can run `nix fmt` to format all files.

**Comment by toastal** [comment]
https://github.com/NixOS/rfcs/pull/166

Sorry. To be clear I was referring to this future Nixpkgs RFC/merge request which looks to to fork `nixfmt`. https://github.com/piegamesde/nixpkgs/tree/nixfmt-2

**Comment by deemp** [comment]
@toastal, I see the implementation isn't merged yet. Could you please create a separate issue that suggests to use the standard formatter once it's ready?

**Comment by toastal** [comment]
No, it’s not merged, but the style the style from the RFC will eventually apply to all of Nixpkgs & by proxy trickle down as a largely defacto style. Deviating from that tends to make just another hurdle to contribution & alejandra’s output is the most dissimilar to what is actively being worked on. Their fork of `nixfmt` will look a bit closer to `nixpkgs-fmt`’s output… which won’t have trailing commas, won’t try to cuddle `let`, space inside inline lists/attrsets, etc. We can see [here](https://github.com/kamadorueda/alejandra/issues/404#issuecomment-1813031210) alejandra will need to wrestle with adopting the style, deprecating themselves, or remaining a niche formatter for those that prefer it to the community standard. Unless its format is truly desired by this package’s maintainer, it doesn’t seem like the wisest choice at present.

**Comment by deemp** [comment]
> Could you please create a separate issue that suggests to use the standard formatter once it's ready?

I've created one (https://github.com/thomashoneyman/purescript-overlay/issues/75).

> Unless its format is truly desired by this package’s maintainer, it doesn’t seem like the wisest choice at present.

I think it's enough for my PR to keep the de-facto used formatting to have relatively small diff.

Hence, if you merge into main another PR that uses a different formatter, I'll rebase my PR on main and format the code using that formatter.

File: generate/lib/src/Utils.purs
Comment (thomashoneyman) [inline generate/lib/src/Utils.purs]:
```suggestion
  liftEffect (Process.exit' 1)

```

File: generate/spago.yaml
Comment (thomashoneyman) [inline generate/spago.yaml]:
```suggestion
        - "unsafe-coerce"

```

File: manifests/build-support/mkDerivationBasic.nix
Comment (thomashoneyman) [inline manifests/build-support/mkDerivationBasic.nix]: Is this record merge necessary if the `meta` function is already creating a meta attr with `mainProgram = pname;`? This looks like a redundant merge to me.

**Review by thomashoneyman — APPROVED** [review]
Looks good to me, thank you! Just a couple comments and, once addressed, we can merge.

File: generate/lib/src/Utils.purs
Comment (deemp) [reply generate/lib/src/Utils.purs]: applied in be88f37

File: generate/spago.yaml
Comment (deemp) [reply generate/spago.yaml]: applied in 4b81bcc268a80ff94b946df46b8169e782227d52

File: manifests/build-support/mkDerivationBasic.nix
Comment (deemp) [reply manifests/build-support/mkDerivationBasic.nix]: The meta function didn't set `mainProgram`. Hence, there was this merge. I feel it's better to set `mainProgram` where `mkDerivationBasic.nix` function is called (09bdc089de03a81825f1a6dae8ec30fafa351aeb) because then,`mainProgram` can be changed independently of `name`.

**Pull request merged**
